### PR TITLE
Lambda  Basic HTTP authentication for resources

### PR DESCRIPTION
**Pull request by phuonghuynh**
For #4

### CODE REVIEW

File: outputs.tf [hunk 1]
```diff
@@ -3,7 +3,9 @@ output "s3_bucket_id" {
 }
 
 output "s3_website_endpoint" {
-  # This doesn't work in terraform due to dependency issues:  # https://github.com/terraform-providers/terraform-provider-aws/issues/1117  # value = "${aws_s3_bucket.main.website_endpoint}"
+  # This doesn't work in terraform due to dependency issues:
```
Comment (ronaldtse) [inline outputs.tf:6]: Apparently now it works, can we use the simple version?

**Comment by ronaldtse** [comment]
@phuonghuynh thank you !

1. Can we move this module (`lambda/*.tf`) to a separate module (new git repo), so that it can be required from this module as optional (not every user of this module needs HTTP AUTH)?

2. ~~In addition, we need to provide a README example on how to use this new lambda auth function.
We need examples for these use cases:~~

* ~~Blacklisting:~~
  1. ~~Protect particular file~~
  2. ~~Protect all files inside a subdirectory~~
  3. ~~Protect all files in the S3 website~~
* ~~Whitelisting:~~
  1. ~~Whitelist publicly accessible files~~

3. ~~Can you also provide in the README (in the new module's README) how to create the "access permissions" file, we can use Terraform's s3_file function to directly create that file with the appropriate content.~~

4. Update this repo's README to indicate how we can use these two modules together.

Thanks!

File: data.tf [hunk 2]
```diff
@@ -1,3 +1,5 @@
 data "aws_region" "main" {
   provider = "aws.main"
 }
+
```
Comment (colindean) [inline data.tf]: Unnecessary whitespace addition

File: lambda.tf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module "lambda" {
+  source = "lambda"
+
+  bucket_name = "${var.bucket_name}"
+  bucket_key = "${var.bucket_key}"
+  basic_user = "${var.basic_user}"
+  basic_password = "${var.basic_password}"
+}
```
Comment (colindean) [inline lambda.tf]: New line at end of file

**Review by colindean — COMMENTED** [review]
Could stand to have `terraform fmt` run on it, too…

**Comment by ronaldtse** [comment]
Seconding @colindean  😉

**Comment by phuonghuynh** [comment]
Good idea @ronaldtse

**Comment by phuonghuynh** [comment]
> Can we move this module (lambda/*.tf) to a separate module (new git repo), so that it can be required from this module as optional (not every user of this module needs HTTP AUTH)?

Lambda Function  won't protect all resources if s3 object not defined (empty or unable to read), so lambda module will be required always. This way, we don't need a separated lambda module (an other GitHub Repo)

**Comment by ronaldtse** [comment]
The reason for separating this to a separate repo, is to allow others that use S3 website but not our current module, to also use the Lambda module. 

Basically, we want the lambda module to be “optional” to our current website module.

**Comment by phuonghuynh** [comment]
@ronaldtse what is the name of the separated Lambda@Edge? "terraform-basic-authenticating-resources-lambda" is good ?

**Comment by ronaldtse** [comment]
How about "terraform-aws-lambda-edge-authentication"?

(https://github.com/riboseinc/terraform-aws-lambda-edge-authentication)

**Comment by phuonghuynh** [comment]
@ronaldtse could you initialize "https://github.com/riboseinc/terraform-aws-lambda-edge-authentication" thus I can fork it ? ^

**Comment by ronaldtse** [comment]
@phuonghuynh you have access :wink:

**Comment by ronaldtse** [comment]
@phuonghuynh are we ready to merge this? Needs a rebase. Thanks!

**Comment by phuonghuynh** [comment]
Yes

File: README.adoc [hunk 4]
```diff
@@ -74,6 +74,17 @@ module "main" {
...
   # Optional WAF Web ACL ID, defaults to none.
   web_acl_id = "${data.terraform_remote_state.site.waf-web-acl-id}"
+
+  # Basic HTTP authentication resources defined in Bucket Object "phuonghqh.booppi.website.htaccess/htaccess.json"
+  # sample "htaccess.json":
+  #   [
+  #     "/sample.png",
+  #     "/private/admin.png"
+  #   ]
+  bucket_name = "phuonghqh.booppi.website.htaccess"
```
Comment (colindean) [inline README.adoc:84]: Looks like from comments on #14 that these aren't actually used? They're not present in `variables.tf`.
